pkg/services/apiserver/builder/openapi.go

```diff
@@ -26,6 +29,92 @@ func GetOpenAPIDefinitions(builders []APIGroupBuilder) common.GetOpenAPIDefiniti
 	}
 }
 
+// Modify the the OpenAPI spec to include the additional routes.
+// Currently this requires: https://github.com/kubernetes/kube-openapi/pull/420
+// In future k8s release, the hook will use Config3 rather than the same hook for both v2 and v3
+func getOpenAPIPostProcessor(builders []APIGroupBuilder) func(*spec3.OpenAPI) (*spec3.OpenAPI, error) {
+	return func(s *spec3.OpenAPI) (*spec3.OpenAPI, error) {
+		if s.Paths == nil {
+			return s, nil
+		}
+		for _, b := range builders {
+			routes := b.GetAPIRoutes()
+			gv := b.GetGroupVersion()
+			prefix := "/apis/" + gv.String() + "/"
+			if s.Paths.Paths[prefix] != nil {
+				copy := spec3.OpenAPI{
+					Version: s.Version,
+					Info: &spec.Info{
+						InfoProps: spec.InfoProps{
+							Title:   gv.String(),
+							Version: setting.BuildVersion,
+						},
+					},
+					Components:   s.Components,
+					ExternalDocs: s.ExternalDocs,
+					Servers:      s.Servers,
+					Paths:        s.Paths,
+				}
+
+				if routes == nil {
+					routes = &APIRoutes{}
+				}
+
+				for _, route := range routes.Root {
+					copy.Paths.Paths[prefix+route.Path] = &spec3.Path{
+						PathProps: *route.Spec,
+					}
+				}
+
+				for _, route := range routes.Namespace {
+					copy.Paths.Paths[prefix+"namespaces/{namespace}/"+route.Path] = &spec3.Path{
+						PathProps: *route.Spec,
+					}
+				}
+
+				// Make the sub-resources (connect) share the same tags as the main resource
+				for path, spec := range copy.Paths.Paths {
+					idx := strings.LastIndex(path, "{name}/")
+					if idx > 0 {
+						parent := copy.Paths.Paths[path[:idx+6]]
+						if parent != nil && parent.Get != nil {
+							for _, op := range GetPathOperations(spec) {
+								if op != nil && op.Extensions != nil {
+									action, ok := op.Extensions.GetString("x-kubernetes-action")
+									if ok && action == "connect" {
+										op.Tags = parent.Get.Tags
+									}
+								}
+							}
+						}
+					}
+				}
+
+				// Support direct manipulation of API results
+				processor, ok := b.(OpenAPIPostProcessor)
```

Comment:
^^ @toddtreece -- is this the best way to support optional features from a builder?

Reply:
checking if the interface is implemented makes sense to me